ami/trainers/image_vae_trainer.py

```diff
@@ -0,0 +1,55 @@
+from functools import partial
+
+import torch
+from torch.distributions import kl_divergence
+from torch.distributions.normal import Normal
+from torch.nn.functional import mse_loss
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader
+
+from ami.data.buffers.buffer_names import BufferNames
+from ami.models.model_names import ModelNames
+from ami.models.vae import VAE, Conv2dDecoder, Conv2dEncoder
+
+from .base_trainer import BaseTrainer
+
+
+class ImageVAETrainer(BaseTrainer):
+    def __init__(
+        self,
+        partial_dataloader: partial[DataLoader[torch.Tensor]],
+        partial_optimizer: partial[Optimizer],
+        device: torch.device,
+    ):
+        super().__init__()
+        self.partial_optimizer = partial_optimizer
+        self.partial_dataloader = partial_dataloader
+        self.device = device
+
+    def on_data_users_dict_attached(self) -> None:
+        self.data_user = self.get_data_user(BufferNames.IMAGE_BUFFER)
+
+    def on_model_wrappers_dict_attached(self) -> None:
+        self.encoder: Conv2dEncoder = self.get_training_model(ModelNames.IMAGE_ENCODER).model
+        self.decoder: Conv2dDecoder = self.get_training_model(ModelNames.IMAGE_DECODER).model
+        self.vae = VAE(self.encoder, self.decoder)
+        self.optimizer_state = self.partial_optimizer(self.vae.parameters()).state_dict()
+
+    def train(self) -> None:
+        optimizer = self.partial_optimizer(self.vae.parameters())
```

Comment:
```suggestion
        self.optimizer_state = self.partial_optimizer(self.vae.parameters()).state_dict()

    def train(self) -> None:
        self.vae = VAE(self.encoder.model, self.decoder.model)
        optimizer = self.partial_optimizer(self.vae.parameters())
```
これは`suggestion`では無いのですが、もしUnwrapされたモデルを用いて `VAE`をコンストラクトしたい場合は、学習時に毎回呼ばれる`train()`メソッド無いでモデルを構築してください。

Reply:
https://github.com/MLShukai/ami/pull/75/commits/23c761b2631da98dfaabbf3a996f71a0466cf53c
optimizer_stateの取得含めて修正しました